src/iterator.cc

```diff
@@ -205,34 +206,26 @@ void checkEndCallback (Iterator* iterator) {
 
 NAN_METHOD(Iterator::Seek) {
   Iterator* iterator = Nan::ObjectWrap::Unwrap<Iterator>(info.This());
-  iterator->GetIterator();
-  leveldb::Iterator* dbIterator = iterator->dbIterator;
-  Nan::Utf8String key(info[0]);
 
-  dbIterator->Seek(*key);
+  leveldb::Slice* target;
+  v8::Local<v8::Object> targetHandle = info[0].As<v8::Object>();
+  LD_STRING_OR_BUFFER_TO_SLICE(_target, targetHandle, target);
+  target = new leveldb::Slice(_target.data(), _target.size());
+
+  // This boolean prevents nexting in Read() and seeking in GetIterator().
   iterator->seeking = true;
+  iterator->GetIterator();
+  iterator->IteratorSeek(target);
+
+  if (iterator->reverse) {
+    // Seek(target) lands on or after the target. But in reverse,
+    // it should land on or *before* the target. I.e. if a seek
+    // to "b" lands on "c", go back to "a".
+    leveldb::Iterator* dbIterator = iterator->dbIterator;
 
-  if (dbIterator->Valid()) {
-    int cmp = dbIterator->key().compare(*key);
-    if (cmp > 0 && iterator->reverse) {
-      dbIterator->Prev();
-    } else if (cmp < 0 && !iterator->reverse) {
-      dbIterator->Next();
-    }
-  } else {
-    if (iterator->reverse) {
-      dbIterator->SeekToLast();
-    } else {
-      dbIterator->SeekToFirst();
-    }
     if (dbIterator->Valid()) {
-      int cmp = dbIterator->key().compare(*key);
-      if (cmp > 0 && iterator->reverse) {
-        dbIterator->SeekToFirst();
+      if (target->compare(dbIterator->key().ToString()) < 0) {
```

Comment:
I'm currently refactoring things so that:

- this bit can be moved to `Iterator::IteratorSeek`, keep all the comparisons there
- and to not perform the actual seek right away, but at the next read